cmd/dlv/main.go

```diff
@@ -49,5 +48,75 @@ func main() {
 		os.Exit(0)
 	}
 
-	cli.Run(os.Args[1:])
+	// Collect launch arguments
+	var processArgs []string
+	var attachPid int
+	switch flag.Args()[0] {
+	case "run":
+		const debugname = "debug"
+		cmd := exec.Command("go", "build", "-o", debugname, "-gcflags", "-N -l")
+		err := cmd.Run()
+		if err != nil {
+			fmt.Errorf("Could not compile program: %s\n", err)
+			os.Exit(1)
+		}
+		defer os.Remove(debugname)
+
+		processArgs = append([]string{"./" + debugname}, flag.Args()...)
+	case "test":
+		wd, err := os.Getwd()
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+		base := filepath.Base(wd)
+		cmd := exec.Command("go", "test", "-c", "-gcflags", "-N -l")
+		err = cmd.Run()
+		if err != nil {
+			fmt.Errorf("Could not compile program: %s\n", err)
+			os.Exit(1)
+		}
+		debugname := "./" + base + ".test"
+		defer os.Remove(debugname)
+
+		processArgs = append([]string{debugname}, flag.Args()...)
+	case "attach":
+		pid, err := strconv.Atoi(flag.Args()[1])
+		if err != nil {
+			fmt.Errorf("Invalid pid: %d", flag.Args()[1])
+			os.Exit(1)
+		}
+		attachPid = pid
+	default:
+		processArgs = flag.Args()
+	}
+
+	// Make a TCP listener
+	listener, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		fmt.Printf("couldn't start listener: %s\n", err)
+		os.Exit(1)
+	}
+	fmt.Printf("started listener on %s\n", listener.Addr())
```

Comment:
This would be better logged instead of printed. Especially for this refactor, the only client will be the CLI.